### DISCUSSION TITLE
Creating implicit commits when `abandon`ing current WCC

### DISCUSSION THREAD

**Discussion by kuchta** [jj-vcs/jj#3671] [Ideas]
I think it's a bad practice to implicitly create news commits when the user doesn't explicitly ask for it.

Possible solution: move WCC to it's parent and in case of multiple parents, randomly choose one and print warning

Relates to `prev`, `next` and possibly others...

**Comment by bnjmnt4n** [comment]
Note that this is the default behavior for all commands: a working-copy commit emptied by a command will always create a new working-copy commit. Perhaps we should create a new flag which avoids creating a new working-copy commit and updates the working copy to be a parent instead?

**Comment by kuchta** [comment]
@bnjmnt4n Yes, it looks like a default behavior, which is quite counter-intuitive and although not destructive, it's something I think should be avoided as possible. I'm personally in favor of the opposite solution, don't make implicitly new commit, but either use some parameter like `--commit`/`-c` for explicitly creating one or preferebly by user defined functions that can utilize multiple commands in sequence, which would make basic command set simpler...

**Reply by bnjmnt4n** [reply]
This would make a command like `jj squash` much harder to use, so I'm not sure if it's advisable to change the default behavior.

**Reply by kuchta** [reply]
You can create an alias, if those 3 keystrokes are too much or a burden for you...

**Reply by kuchta** [reply]
Because hardwiring behavior for custom workflows goes against that orthogonality principle which I thought you are striving for...

**Reply by kuchta** [reply]
Or you can make default command arguments configurable or make aliases more powerful so that they can overshadow basic commands both of which could be a quite problematic, though...

**Reply by bnjmnt4n** [reply]
One of the features of jj is that commands can be applied consistently to any revision, without special handling of the working-copy commit. Here's how the current behavior is consistent:

`jj squash -r A`, where A is not the working copy: A is squashed into its parent commit; further edits to the current working copy do not affect A's parent commit
`jj squash -r @`: @ is squashed into its parent commit; a new working-copy commit is created so further edits to the current working copy do not affect the parent commit

`jj abandon -r A`, where A is not the working copy: A is abandoned; further edits to the current working copy do not affect A's parent commit
`jj abandon @`: @ is abandoned; a new working-copy commit is created so further edits to the current working copy do not affect the original `@`'s parent commit

With your proposed change, the commit being modified with further edits to the working copy is inconsistent:

`jj squash -r A`, where A is not the working copy: A is squashed into its parent commit; further edits to the current working copy do not affect A's parent commit
`jj squash -r @`: @ is squashed into its parent commit; further edits to the current working copy modify the original `@`'s parent

`jj abandon A`, where A is not the working copy: A is abandoned; further edits to the current working copy do not affect A's parent commit
`jj abandon @`: @ is abandoned; further edits to the current working copy modify the original `@`'s parent commit

> I think it's a bad practice to implicitly create news commits when the user doesn't explicitly ask for it.

Any reason why you would think so? Jujutsu also has the implicit behavior of modifying the existing commit when the working copy is changed, which is the primary difference between it and other VCSes. As long as the behavior is consistent (which it currently is, any emptied working copy commit due to a command will create a new empty working copy commit), I don't see why it's a bad practice.

> Because hardwiring behavior for custom workflows goes against that orthogonality principle which I thought you are striving for...

Could you elaborate more on what is the orthogonality principle? I don't think I've seen it mentioned in the past discussions.

**Reply by kuchta** [reply]
I didn't say current behavior is not self-consistent, I just say it's counter-intuitive and has at first sight non-obvious conditonal logic for polluting commit graph that is workflow specific.

As you say "Jujutsu also has the implicit behavior of modifying the existing commit when the working copy is changed" so probably no one minimally familiar with it would go and make modifications to the WC, if that's not something they will want to do and printing WCC after any relevant command makes that quite obvious.

You asked for a reason it's a bad practice to implicitly create new commits. Apart from the fact that it's counter-intuitive (which Martin confirmed by saying that users are confused by commands like `next`/`prev` creating new commits), it's generally good engineering practice to build simple constrained and predictable functions (KISS) that does one thing well and combine such functions to build more complex functions/scenarios/workflows. User facing conditional logic like the one creating new commits goes against that, IMHO...

Orthogonality principle (which I encountered somewhere mentioned, but can't find it now, so I could be just in some third party tutorial) is just that, building a system with simple non-overlapping functions that does one specific thing well and combining such functions to build more complex ones (system/workflow/etc)...

**Reply by kuchta** [reply]
And I even don't say jj coudn't have some idiomatic way how to do things, just that it's better done using some default set of user configurable aliases/functions than hardcoded in the basic commands. One thing that git did quite well (apart from it's data model) even though probably not quite intentionally was to come up with that plumbing/porcelain command separation and here it could be even improved by making that separtion visible and porcelain fully configurable...

**Reply by bnjmnt4n** [reply]
> Apart from the fact that it's counter-intuitive

Your claim is that creating new commits is counter-intuitive because the user didn't explicitly create a new commit and it can confuse users. You suggest that instead of setting the working-copy to a newly created commit, you update the working copy to the previous parent of an abandoned commit (even randomly choosing a parent if there's more than one). Wouldn't that also be counter-intuitive, since the user didn't ask to explicitly change the working-copy?

One can argue that both those behaviors are counter-intuitive, and maybe Jujutsu should just error if you're trying to abandon the current working copy. However, IMO, that isn't a good default: because the user has explicitly requested to abandon a commit, you also shouldn't be disallowed from doing so just because your working copy is set to that commit.

Thus, Jujutsu should default to a solution which best matches the user's intent:
- The user did not request to switch the working copy commit to any parent commit, so the working copy commit should not be changed to the parent commit (especially since this could cause unintentional modification of the commit).
- Does the user want to create a new commit? Maybe, or maybe not. However, it is safer to create a new empty commit than to allow the user to modify the parent commit unintentionally. Hence, I'd say this is the better default option.

Hence, I recommend that the default should still be to create new working copy commits implicitly, and add a flag to disable that if not desired.

(I also couldn't find any other requests to change this behavior, so I would say that the current behavior is not as un-intuitive as you claim.)

> Apart from the fact that it's counter-intuitive (which Martin confirmed by saying that users are confused by commands like next/prev creating new commits), it's generally good engineering practice to build simple constrained and predictable functions (KISS) that does one thing well and combine such functions to build more complex functions/scenarios/workflows. User facing conditional logic like the one creating new commits goes against that, IMHO...

Both the current behavior and your suggested change has user facing conditional logic: the command will update the working copy commit only if the working copy commit was `abandoned`/made empty, otherwise it won't modify the working copy. So I'm not sure what your point is here.

**Reply by bnjmnt4n** [reply]
> You can create an alias, if those 3 keystrokes are too much or a burden for you...

I'm not sure why you'd want to keep making personal statements like these, because they don't further your argument in any way, and you're more likely to be regarded as acting in bad faith.

**Reply by kuchta** [reply]
The point is, if you abandon WCC then it's only logical you HAVE TO change it to something and so the user will expect that to happen, so it's intuitive, but creating new commit when it's not neccesary is something user wouldn't expect, so it's unintuitive. Would you like your editor to create new files in the file system if you tell him to delete current one? Probably not 😉
Most of the time there would be just one parent, so changing to it is quite intuitive and deterministic behavior (where else would you want to change?) so it's not deterministic only in case of multiple parents and that's why warning should be issued. I understand why you think it's not safe to move WCC to parent, but it's not if it's the only logical thing the system can do at that time and you would know that for sure because that command will print WCC and it's metadata at the end and so would do the GUI. Users will learn that to prevent further modifications to finished WC commit, thay have to do `jj new` after such operation, if that's what they want. Any system that thinks it knows better what the user wants than the user himself is probably doomed to extinction 🙂. Users of VCSs are not mindless idiots, but usually technicaly oriented people so if this is default behavior of the system and not just some command specific one, they will know, don't worry 😉 The system has some notion of immutable revisions, so it will probably prevent you (or should do so) to easily change a revision that is already shared with others...

**Reply by bnjmnt4n** [reply]
> Would you like your editor to create new files in the file system if you tell him to delete current one? Probably not 😉

No, but I also wouldn't expect it to open the previously edited file, and allow me to continue editing it (hoping that I would read the warning).

> it's only logical you HAVE TO change it to something and so the user will expect that to happen, so it's intuitive

As mentioned above, an alternative would be to error if the commit to abandon is the working copy commit.

> but creating new commit when it's not neccesary is something user wouldn't expect, so it's un-intuitive.

Personally, I would prefer consistency no matter what revision is operated on, which doesn't involve conditionally changing the working copy to a parent. Additionally, I don't think there are many other complaints about this, so perhaps most other users do not share the same opinion that this behavior is un-intuitive.

This is why my proposal is to create a new flag which allows you to disallow creating new commits implicitly. If this is the desired behavior of most users, users will start to use this flag, and there will be more feedback about changing the default.

**Reply by martinvonz** [reply]
> Apart from the fact that it's counter-intuitive (which Martin confirmed by saying that users are confused by commands like `next`/`prev` creating new commits)

Clarification: I said that people are confused specifically by `jj next/prev`. I think the confusion there is that the movement is relative to the *parent* of the working copy, not relative to the working copy itself (which is what you get with `--edit`).

**Reply by martinvonz** [reply]
> * However, it is safer to create a new empty commit than to allow the user to modify the parent commit unintentionally.

Especially if the parent commit is immutable (issue #3201 is relevant).

**Reply by kuchta** [reply]
Aha, okay. Yes, that's even more confusing 🫤 Thanks for clarification 👍

**Reply by martinvonz** [reply]
> Would you like your editor to create new files in the file system if you tell him to delete current one? Probably not 😉

I think your proposal is analogous to having the editor instead edit a file in the parent directory (only kind of - it would be closer to say that you would edit the parent directory but you can't typically edit directories in editors). I'm not sure that's better behavior. But let's not get distracted by analogies.

That said, I was going to say that I think another reasonable behavior is to leave the working copy on the abandoned commit, which - incidentally - is quite similar to what shell typically do when you delete the current working directory :) That behavior makes sense to me, but it also seems less useful. I very often use `jj abandon` to abandon my current work. Sure, I could create an alias for `jj abandon && jj new @-` for myself, but if we think that almost every user would want that alias, then I think it's better that we just make the command do that.

**Reply by kuchta** [reply]
Quote reply not working, so let's quote that manually 🙂

> I think your proposal is analogous to having the editor instead edit a file in the parent directory (only kind of - it would be closer to say that you would edit the parent directory but you can't typically edit directories in editors). I'm not sure that's better behavior. But let's not get distracted by analogies.

Yes, that would be probably even better analogy (and some unix systems even allow you to edit directories in the editor) and I would be quite fine with such behavior

> That said, I was going to say that I think another reasonable behavior is to leave the working copy on the abandoned commit, which - incidentally - is quite similar to what shell typically do when you delete the current working directory :) That behavior makes sense to me, but it also seems less useful. 

That behavior is the result of handling inodes in the unix kernel (which just removes the directory entry, but not the inode itself until the program exits) and in some way it's better than that of the windows e.g. because running programs can continue without error. On the other hand, you would possibly loose data after exiting, so probably not a good model for something that should retain data as much as possible. If you would want to replicate such behavior in jj, it would probably mean setting parent commit to none and letting (future) GC to collect it. But I don't know if abandon does not already do something like it, though...

> I very often use jj abandon to abandon my current work. Sure, I could create an alias for jj abandon && jj new @- for myself, but if we think that almost every user would want that alias, then I think it's better that we just make the command do that.

Regarding `jj abandon`, couldn't you just use `jj restore` for that? You will retain metadata, but I suppose people usually set description when they are finished...

**Reply by martinvonz** [reply]
> Regarding `jj abandon`, couldn't you just use `jj restore` for that? You will retain metadata, but I suppose people usually set description when they are finished...

Yes, that would usually work well enough. I would switch to that if we changed the behavior to not update to a parent when the working copy is abandoned.

**Reply by kuchta** [reply]
Well, I'm now quite interested what actually happens when you abandon a commit? Does jj delete it, or set the parent to none? How would jj in the future implement GC, if it's not based on branches? In branch model it's simple, whatever you can't follow backwards to root is collected, but in a branchless model... Would you take opposite approach of following root commit forward and collecting everying not found?

**Reply by ilyagr** [reply]
> Well, I'm now quite interested what actually happens when you abandon a commit? 

Whether a commit is visible is an extra piece of metadata stored in jj's store (outside the part that's managed by git, same place where the change id of a commit is remembered). There are also some rules that should be enforced by `jj`: the working copy is always visible, ancestors of visible commits are also visible.

**Reply by kuchta** [reply]
Thank you @ilyagr for clarification. And the default is that everything is visible until abandoned? Can actually some ancestor of a visible commit be invisible if abandon rebases it's target descendants onto it's parent?

**Reply by ilyagr** [reply]
The primary ways commits become hidden is when you edit them or rebase them.

As the tutorial explains, any change to the working tree changes the commit id of the current change. This means that the old commit became hidden and a new visible commit was created with the same change id.

Similarly, when you rebase a commit, the old version becomes hidden, and a new visible commit with the same change id is created. 

You can see all the hidden "previous versions" of commits by running `jj obslog`.

`jj abandon` hides a commit without creating a replacement visible commit. Then, a rebase automatically happens, but you can think of this as a separate operation. (Strictly speaking, for the rules I described to always hold, the rebase would logically have to happen first, so that the commit doesn't have any visible descendants when it becomes hidden)

**Reply by kuchta** [reply]
Oh of course, I didn't realize that, thank you, @ilyagr 🙏🏻

**Comment by kuchta** [comment]
> No, but I also wouldn't expect it to open the previously edited file, and allow me to continue editing it (hoping that I would read the warning).
> 
> > it's only logical you HAVE TO change it to something and so the user will expect that to happen, so it's intuitive
> 
> As mentioned above, an alternative would be to error if the commit to abandon is the working copy commit.

Okay, I agree. That is probably even better default, especially when you can do `jj restore` to achieve basicaly the same thing

> > but creating new commit when it's not neccesary is something user wouldn't expect, so it's un-intuitive.
> 
> Personally, I would prefer consistency no matter what revision is operated on, which doesn't involve conditionally changing the working copy to a parent. Additionally, I don't think there are many other complaints about this, so perhaps most other users do not share the same opinion that this behavior is un-intuitive.

Many people are not oriented towards evolution/development and just accept suboptimal status quo (and some are even able to fight for it) 🙂

> This is why my proposal is to create a new flag which allows you to disallow creating new commits implicitly. If this is the desired behavior of most users, users will start to use this flag, and there will be more feedback about changing the default.

Yes, why do the right thing, when you can do wrong one and wait for somebody to yell at you  😀

Maybe jj should ask (or make that configurable) if it should update the current commit, or move the changes to the new one, but implicitly creating new commits is not a good default the same way as you wouldn't want your editor to behave that way...

**Reply by kuchta** [reply]
Well, when I think about it, maybe it's not a bad idea to make that configurable, since if there would be some watch mode (or editor extension that watches for changes) in the future that would create new commit after each save, that could also be interesting workflow for some projects...
